packages/flutter/lib/src/services/image_cache.dart

```diff
@@ -11,25 +12,55 @@ import 'fetch.dart';
 import 'image_decoder.dart';
 import 'image_resource.dart';
 
-Future<ui.Image> _fetchImage(String url) async {
-  UrlResponse response = await fetchUrl(url);
-  if (response.statusCode >= 400) {
-    print("Failed (${response.statusCode}) to load image $url");
-    return null;
+/// Implements a way to retrieve an image, for example by fetching it from the network.
+/// Also used as a key in the image cache.
+abstract class ImageProvider {
+  Future<ui.Image> loadImage();
+}
+
+class _UrlFetcher implements ImageProvider {
+  final String _url;
+
+  _UrlFetcher(this._url);
+
+  Future<ui.Image> loadImage() async {
+    UrlResponse response = await fetchUrl(_url);
+    if (response.statusCode >= 400) {
+      print("Failed (${response.statusCode}) to load image $_url");
+      return null;
+    }
+    return await decodeImageFromDataPipe(response.body);
+  }
+
+  bool operator ==(o) => o is _UrlFetcher && _url == o._url;
+  int get hashCode => _url.hashCode;
+}
+
+class RawImageProvider implements ImageProvider {
+  final Uint8List bytes;
+
+  RawImageProvider(this.bytes);
```

Comment:
Assuming we don't want to calculate hash and override equality for Uint8List for performance reasons, IMO we should optionally take a 'String uniqueKey' and use that for equality/hashCode when provided. This way applications can control the cache better and gain better hit rates if their logic allows them to map the bytes to some unique identifier based on their own custom logic. If key is not provided, IMO using identity of the Uint8List rather than RawImageProvider might cause more cache hits.